libbeat/reader/readfile/timeout.go

```diff
@@ -75,14 +76,21 @@ func (r *TimeoutReader) Next() (reader.Message, error) {
 			}
 		}()
 	}
-
+	if r.timer == nil {
+		r.timer = time.NewTimer(r.timeout)
+	} else {
+		r.timer.Reset(r.timeout)
```

Comment:
Unless we're seeing noticeable memory / cpu overhead from it, I actually like the ephemeral version better (allocate a timer each time). Reusing timers is a lot more subtle -- because `Reset` can only be called on a timer that has stopped or expired, and `<-timer.C` can deadlock if it's called in the wrong state... it looks to me like you've handled all the cases correctly here, but it takes considerable thought to make sure.

If this is a significant performance fix over the previous version, then this looks ok, but please comment on these issues (e.g. pointing out that `timer.Reset` is safe because in every code path the timer is `Stop`ped or expired before this function returns, and that `<-r.timer.C` can't deadlock because `Stop` can only be called once per `Reset` and thus a false return value means there is an expiration signal waiting...)

If this doesn't affect performance much, though, I'd rather just create and deallocate the value every time, to save the next person who reads this function the cognitive load of thinking about state invariants :-)